lib/ansible/modules/web_infrastructure/dyn_record.py

```diff
@@ -0,0 +1,562 @@
+        foundrecord = self.recordget(node, record_type, rdata=rdata)
+
+        if foundrecord is None:
+            return {'deleted': False}
+        else:
+            try:
+                response, info = fetch_url(self.module, self.session.api_uri + "/REST/" +
+                                           str(record_type).upper() + "Record/" +
+                                           str(self.zone) + "/" + str(node) + "." +
+                                           str(self.zone) + "/" +
+                                           str(foundrecord['record_id']),
+                                           method="delete",
+                                           headers={'Content-Type': 'application/json', 'Auth-Token': self.session.auth_token})
+            except Exception as exception:
+                self.module.fail_json(msg=str(exception))
+            else:
+                if info["status"] >= 400:
+                    self.module.fail_json(msg=str(info))
+                else:
+                    # body = response.read()
+                    return {'deleted': True}
+
+    def getpendingchanges(self):
+        '''Get a list of the zone pending changes.'''
+        try:
+            response, info = fetch_url(self.module, self.session.api_uri + "/REST/ZoneChanges/" +
+                                       str(self.zone) + "/",
+                                       method="get",
+                                       headers={'Content-Type': 'application/json', 'Auth-Token': self.session.auth_token})
+        except Exception as exception:
+            self.module.fail_json(msg=str(exception))
+        else:
+            if info["status"] >= 400:
+                self.module.fail_json(msg=str(info))
+            else:
+                body = response.read()
+                zonechanges = json.loads(body)['data']
+                return zonechanges
+
+    def publish(self):
+        '''Publish pending zone changes'''
+        payload = {'publish': True, 'notes': str(self.module.params['notes'])}
+
+        try:
+            response, info = fetch_url(self.module, self.session.api_uri + "/REST/Zone/" +
+                                       str(self.zone) + "/",
+                                       method="put",
+                                       data=self.module.jsonify(payload),
+                                       headers={'Content-Type': 'application/json', 'Auth-Token': self.session.auth_token})
+        except Exception as exception:
+            self.module.fail_json(msg=str(exception))
+        else:
+            if info["status"] >= 400:
+                self.module.fail_json(msg=str(info))
+            else:
+                body = response.read()
+                return json.loads(body)
+
+
+def main():
+    """Initiates module."""
+    module = AnsibleModule(
+        argument_spec=dict(
+            dyn_customer_name=dict(required=False, default=None, type='str'),
```

Comment:
instead of the os env handling code in the class DynectAPI, I would just handle it here, e.g. for `dyn_customer_name`, the only thing in DynectAPI is to check for `None`.
~~~diff
- dyn_customer_name=dict(required=False, default=None, type='str'),
+ dyn_customer_name=dict(required=False, default=os.environ.get('DYN_CUSTOMER_NAME'), type='str'),
~~~

Reply:
Hey @resmo ! Glad to have you onboard :), let me answer your questions :
> I noticed you pass a list of records, but in the code you loop over the list to make the api call. So in my point of view, there is no much gain to pass a list and I would change it to only handle one record at a time


This is indeed a choice I made to reduce API calls not to trigger Dyn's API rate limiting:
- As you noticed, for the authentication part first.
- For the future, to be able to handle multiple record duplicates present in the parameters and/or upstream on the DNS zone. By duplicate records, I mean records that share the same fqdn for specific record types like TXT, but not the same value. Handling only a single record at a time would not get us the expected results (i.e. "which record among all these would be updated?").
Therefore, I would like to keep things that way, but we could always add more optional input parameters ("node", "ttl", "rdata") to be able to answer this simpler need if the "records" parameter is not present, what do you think?

Dyn exposes endpoints to handle this particular update case which this PR does not handle yet. 
I was working on a complete rework with a colleague of mine, but since this immediate need has vanished for our use case, I thought I would make a PR of the existing code.

> You already separated out the authentication part, I assume when more dyn modules arrive, we should reuse it and move it to module_utils to share it across the dyn modules.

:+1: , this was mainly to be able to import the module within a custom library path as long as the PR isn't merged, I couldn't seem to specify a "library_utils" path according to the docs! I'll get right to it.

> Speaking of more dyn modules, what other modules are you working on currently?

None so far. Though I think I'll get to working on a traffic director module in the following weeks (probably january 2018) if you do not first ;)

> Further, the path the module is currently located is not ideal. Please move it to to a subpath modules/net_tools/dynect (you have to create an emtpy init.py file).

Ok will do.

Thanks for your much appreciated feedback!

Reply:
Ok i'll get to it